go/client/cas_test.go

```diff
@@ -625,3 +628,158 @@ func TestFlattenActionOutputs(t *testing.T) {
 		}
 	}
 }
+
+func TestDownloadActionOutputs(t *testing.T) {
+	ctx := context.Background()
+	listener, err := net.Listen("tcp", ":0")
+	if err != nil {
+		t.Fatalf("Cannot listen: %v", err)
+	}
+	defer listener.Close()
+	server := grpc.NewServer()
+	fake := fakes.NewCAS()
+	bsgrpc.RegisterByteStreamServer(server, fake)
+	regrpc.RegisterContentAddressableStorageServer(server, fake)
+	go server.Serve(listener)
+	defer server.Stop()
+	c, err := client.Dial(ctx, instance, client.DialParams{
+		Service:    listener.Addr().String(),
+		NoSecurity: true,
+	}, client.UseBatchOps(true))
+	if err != nil {
+		t.Fatalf("Error connecting to server: %v", err)
+	}
+	defer c.Close()
+
+	fooDigest := fake.Put([]byte("foo"))
+	barDigest := fake.Put([]byte("bar"))
+	dirB := &repb.Directory{
+		Files: []*repb.FileNode{
+			{Name: "foo", Digest: fooDigest.ToProto(), IsExecutable: true},
+		},
+	}
+	bDigest := digest.TestNewFromMessage(dirB)
+	dirA := &repb.Directory{
+		Directories: []*repb.DirectoryNode{
+			{Name: "b", Digest: bDigest.ToProto()},
+		},
+		Files: []*repb.FileNode{
+			{Name: "bar", Digest: barDigest.ToProto()},
+		},
+	}
+	aDigest := digest.TestNewFromMessage(dirA)
+	root := &repb.Directory{
+		Directories: []*repb.DirectoryNode{
+			{Name: "a", Digest: aDigest.ToProto()},
+			{Name: "b", Digest: bDigest.ToProto()},
+		},
+	}
+	tree := &repb.Tree{
+		Root:     root,
+		Children: []*repb.Directory{dirA, dirB},
+	}
+	treeBlob, err := proto.Marshal(tree)
+	if err != nil {
+		t.Errorf("failed marshalling Tree: %s", err)
+	}
+	treeA := &repb.Tree{
+		Root:     dirA,
+		Children: []*repb.Directory{dirB},
+	}
+	treeABlob, err := proto.Marshal(treeA)
+	if err != nil {
+		t.Errorf("failed marshalling Tree: %s", err)
+	}
+	treeDigest := fake.Put(treeBlob)
+	treeADigest := fake.Put(treeABlob)
+	ar := &repb.ActionResult{
+		OutputFiles: []*repb.OutputFile{
+			&repb.OutputFile{Path: "foo", Digest: fooDigest.ToProto()}},
+		OutputFileSymlinks: []*repb.OutputSymlink{
+			&repb.OutputSymlink{Path: "x/bar", Target: "../dir/a/bar"}},
+		OutputDirectorySymlinks: []*repb.OutputSymlink{
+			&repb.OutputSymlink{Path: "x/a", Target: "../dir/a"}},
+		OutputDirectories: []*repb.OutputDirectory{
+			&repb.OutputDirectory{Path: "dir", TreeDigest: treeDigest.ToProto()},
+			&repb.OutputDirectory{Path: "dir2", TreeDigest: treeADigest.ToProto()},
+		},
+	}
+	execRoot, err := ioutil.TempDir("", "DownloadOuts")
+	if err != nil {
+		t.Fatalf("failed to make temp dir: %v", err)
+	}
+	defer os.RemoveAll(execRoot)
+	err = c.DownloadActionOutputs(ctx, ar, execRoot)
+	if err != nil {
+		t.Fatalf("error in DownloadActionOutputs: %s", err)
```

Comment:
Errorf since this is what we test?

Reply:
Hmm, yes, to be consistent with other things, but this is a good example of why I disagree with the pattern. If you have a 

res1, res2, ..., err := ComplexFunction()

and you do a Errorf instead of a Fatalf on a err != nil check, you will get a huge amount of test errors that are all caused by an actual single error. So only the first one will actually matter, you'd need to find the first one and ignore the rest. I agree that you can test res1...resK individually with Errorf, but the err != nil check seems gating to everything else.

Reply:
I think the "keep going" motivation is to print all the errors related to this call at once, so that you fix it at once. Note that if you Fatal on err != nil, you wouldn't notice if you have a bug where you return an error and non-nil outputs. With Errorf you verify the entire output of the function (including the error).